### PR TITLE
feat(plugin config): add dependency for delete plugin config

### PR DESCRIPTION
**Pull request by zaunist**
### What this PR does / why we need it:
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
fix https://github.com/apache/apisix/issues/5917

When a plugin config is referencing by any route, it should not be deleted. 

example:

create plugin config:
```
curl http://127.0.0.1:9180/apisix/admin/plugin_configs/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -i -d '
{                                                                                                                                                                                                         
    "desc": "吾乃插件配置1",                                                                                                                                                                              
    "plugins": {                                                                                                                                                                                          
        "limit-count": {                                                                                                                                                                                  
            "count": 2,                                                                                                                                                                                   
            "time_window": 60,                                                                                                                                                                            
            "rejected_code": 503                                                                                                                                                                          
        }                                                                                            
    }                                                                                                                                                                                                     
}' 
```
bounding route to plugin config:
```
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -i -d '
{
    "uris": ["/index.html"],
    "plugin_config_id": 1,
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "39.97.63.215:80": 1
        }
    }
}'
```
Then, delete plugin config, apisix will return 400 error.
```
▶ curl -X DELETE http://127.0.0.1:9080/apisix/admin/plugin_configs/1  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1'
{"error_msg":"can not delete this plugin config, route [1] is still using it now"}
```

### Pre-submission checklist:

<!--
Please follow the PR manners:
1. Use Draft if the PR is not ready to be reviewed
2. Test is required for the feat/fix PR, unless you have a good reason
3. Doc is required for the feat PR
4. Use a new commit to resolve review instead of `push -f`
5. If you need to resolve merge conflicts after the PR is reviewed, please merge master but do not rebase
6. Use "request review" to notify the reviewer once you have resolved the review
7. Only reviewer can click "Resolve conversation" to mark the reviewer's review resolved
-->

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by zaunist** [comment]
cc @tzssangglass

File: t/admin/delete-plugin-configs.t [hunk 1]
```diff
@@ -0,0 +1,174 @@
+#
```
Comment (spacewander) [inline t/admin/delete-plugin-configs.t:1]: Could we add it to `t/admin/plugin-configs.t`?

File: t/admin/delete-plugin-configs.t [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                    "plugins": {
+                        "limit-count": {
+                            "count": 2,
+                            "time_window": 60,
+                            "rejected_code": 503,
+                            "key": "remote_addr"
+                        }
+                    }
+                }]],
+                [[{
```
Comment (spacewander) [inline t/admin/delete-plugin-configs.t:44]: We don't need to check the response in every test.

**Review by tokers — APPROVED** [review]

File: apisix/admin/plugin_config.lua [hunk 3]
```diff
@@ -100,13 +102,33 @@ end
 
 
 function _M.delete(id)
+    if not id then
+        return 400, {error_msg = "missing plugin config id"}
+    end
+
+    local routes, routes_ver = get_routes()
+    core.log.info("routes: ", core.json.delay_encode(routes, true))
```
Comment (spacewander) [inline apisix/admin/plugin_config.lua:110]: Is it necessary to dump all the routes?

File: t/admin/delete-plugin-configs.t (hunk 1 above)
Comment (spacewander) [reply t/admin/delete-plugin-configs.t:1]: We should not add more new `delete-*` files. For example, the new case in https://github.com/apache/apisix/pull/4575 is added in `proto3.t`.
I also submit a PR to remove the stale cases: https://github.com/apache/apisix/pull/6121

Comment (zaunist) [reply t/admin/delete-plugin-configs.t:1]: Got it.

File: apisix/admin/plugin_config.lua (hunk 3 above)
Comment (zaunist) [reply apisix/admin/plugin_config.lua:110]: I did it with reference to the processing of upstream. At the same time, I also considered whether to logging all route information. If it is not log here, does it also need to be modified in upstream?
https://github.com/apache/apisix/blob/c38e94f20144b0ba01018970fa954c935a325ac1/apisix/admin/upstreams.lua#L130

File: t/admin/delete-plugin-configs.t (hunk 2 above)
Comment (zaunist) [reply t/admin/delete-plugin-configs.t:44]: Got it.

**Review by tzssangglass — APPROVED** [review]

File: t/admin/plugin-configs.t [hunk 4]
```diff
@@ -430,3 +430,139 @@ passed
...
+                        "type": "roundrobin"
+                    },
+                }]]
+                )
+
+            ngx.status = code
+            ngx.say(body)
+        }
+    }
+--- request
```
Comment (spacewander) [inline t/admin/plugin-configs.t:494]: We can remove the duplicate sections which are set in the beginning.

File: t/admin/plugin-configs.t [hunk 5]
```diff
@@ -430,3 +430,139 @@ passed
...
+GET /t
+--- response_body
+passed
+--- no_error_log
```
Comment (spacewander) [inline t/admin/plugin-configs.t:498]: Ditto

File: t/admin/plugin-configs.t [hunk 6]
```diff
@@ -430,3 +430,124 @@ passed
...
+--- response_body
+passed
+
+
+
+=== TEST 12: delete-plugin configs failed(id: 1)
+--- config
+    location /t {
+        content_by_lua_block {
+            ngx.sleep(0.3)
```
Comment (bisakhmondal) [inline t/admin/plugin-configs.t:503]: Hi, is there any reason for which we need to add this 0.3-sec sleep?
Thanks.

File: t/admin/plugin-configs.t [hunk 7]
```diff
@@ -430,3 +430,124 @@ passed
...
+{"error_msg":"can not delete this plugin config, route [1] is still using it now"}
+--- error_code: 400
+
+
+
+=== TEST 13: delete route(id: 1)
+--- config
+    location /t {
+        content_by_lua_block {
+            ngx.sleep(0.3)
```
Comment (bisakhmondal) [inline t/admin/plugin-configs.t:523]: ditto

File: t/admin/plugin-configs.t [hunk 8]
```diff
@@ -430,3 +430,124 @@ passed
...
+--- response_body
+passed
+
+
+
+=== TEST 14: delete plugin-configs(id: 1)
+--- config
+    location /t {
+        content_by_lua_block {
+            ngx.sleep(0.3)
```
Comment (bisakhmondal) [inline t/admin/plugin-configs.t:542]: ditto

**Review by bisakhmondal — APPROVED** [review]
lgtm%nits

File: t/admin/plugin-configs.t (hunk 6 above)
Comment (zaunist) [reply t/admin/plugin-configs.t:504]: The current step depends on the previous steps, so set a short time for etcd to sync data.

**Comment by spacewander** [comment]
I found the CI is broken with msg:
```
nginx: [emerg] "listen" directive is not allowed here in /home/runner/work/apisix/apisix/t/servroot/conf/nginx.conf:202
```
https://github.com/apache/apisix/runs/4836910403?check_suite_focus=true

**Comment by zaunist** [comment]
> I found the CI is broken with msg:
> 
> ```
> nginx: [emerg] "listen" directive is not allowed here in /home/runner/work/apisix/apisix/t/servroot/conf/nginx.conf:202
> ```
> 
> https://github.com/apache/apisix/runs/4836910403?check_suite_focus=true

Thanks for your careful, but I don't know how to slove this problem. Any one can help me to do this ?

**Comment by spacewander** [comment]
Usually, it is because of a malformed config section. Could you check the `t/servroot/conf/nginx.conf`?

**Comment by zaunist** [comment]
> Usually, it is because of a malformed config section. Could you check the `t/servroot/conf/nginx.conf`?

Got it, and CI passed now.

**Review by spacewander — APPROVED** [review]

**Pull request merged**